sdks/go/pkg/beam/runners/prism/internal/worker/worker.go

```diff
@@ -0,0 +1,424 @@
+				slog.Debug("StateRequest_Get", prototext.Format(req), "bundle", b)
+
+				var data [][]byte
+				switch key.GetType().(type) {
+				case *fnpb.StateKey_IterableSideInput_:
+					ikey := key.GetIterableSideInput()
+					wKey := ikey.GetWindow()
+					var w typex.Window
+					if len(wKey) == 0 {
+						w = window.GlobalWindow{}
+					} else {
+						w, err = exec.MakeWindowDecoder(coder.NewIntervalWindow()).DecodeSingle(bytes.NewBuffer(wKey))
+						if err != nil {
+							panic(fmt.Sprintf("error decoding iterable side input window key %v: %v", wKey, err))
+						}
+					}
+					winMap := b.IterableSideInputData[ikey.GetTransformId()][ikey.GetSideInputId()]
+					var wins []typex.Window
+					for w := range winMap {
+						wins = append(wins, w)
+					}
+					slog.Debug(fmt.Sprintf("side input[%v][%v] I Key: %v Windows: %v", req.GetId(), req.GetInstructionId(), w, wins))
+					data = winMap[w]
+
+				case *fnpb.StateKey_MultimapSideInput_:
+					mmkey := key.GetMultimapSideInput()
+					wKey := mmkey.GetWindow()
+					var w typex.Window
+					if len(wKey) == 0 {
+						w = window.GlobalWindow{}
+					} else {
+						w, err = exec.MakeWindowDecoder(coder.NewIntervalWindow()).DecodeSingle(bytes.NewBuffer(wKey))
+						if err != nil {
+							panic(fmt.Sprintf("error decoding iterable side input window key %v: %v", wKey, err))
+						}
+					}
+					dKey := mmkey.GetKey()
+					winMap := b.MultiMapSideInputData[mmkey.GetTransformId()][mmkey.GetSideInputId()]
+					var wins []typex.Window
+					for w := range winMap {
+						wins = append(wins, w)
+					}
+					slog.Debug(fmt.Sprintf("side input[%v][%v] MM Key: %v Windows: %v", req.GetId(), req.GetInstructionId(), w, wins))
+
+					data = winMap[w][string(dKey)]
+
+				default:
+					panic(fmt.Sprintf("unsupported StateKey Access type: %T: %v", key.GetType(), prototext.Format(key)))
+				}
+
+				// Encode the runner iterable (no length, just consecutive elements), and send it out.
+				// This is also where we can handle things like State Backed Iterables.
+				var buf bytes.Buffer
+				for _, value := range data {
+					buf.Write(value)
+				}
+				responses <- &fnpb.StateResponse{
+					Id: req.GetId(),
+					Response: &fnpb.StateResponse_Get{
+						Get: &fnpb.StateGetResponse{
+							Data: buf.Bytes(),
+						},
+					},
+				}
+			default:
+				panic(fmt.Sprintf("unsupported StateRequest kind %T: %v", req.GetRequest(), prototext.Format(req)))
+			}
+		}
+	}()
+	for resp := range responses {
+		if err := state.Send(resp); err != nil {
+			slog.Error("state.Send error", err)
+		}
+	}
+	return nil
+}
+
+// DataService is slated to be deleted in favour of stage based state
+// management for side inputs.
+type DataService struct {
+	// TODO actually quick process the data to windows here as well.
+	raw map[string][][]byte
+}
+
+// Commit tentative data to the datastore.
+func (d *DataService) Commit(tent engine.TentativeData) {
+	if d.raw == nil {
+		d.raw = map[string][][]byte{}
```

Comment:
Just curious to learn the motivation to initialize the map here (same for `engine.TentativeData`) vs in a `New` function as is done e.g. for the maps in `worker.W`

Reply:
Partly due to an earlier iteration on the design.

Generally, it's a [Go Proverb](https://go-proverbs.github.io/) it's ideal to [Make the Zero Value useful](https://www.youtube.com/watch?v=PAAkCSZUG1c&t=385s), especially when others are expecting to use the type. It simplifies initialization of the type, in particularly in larger data structures.

It's less critical when it's scope is only in a single package or project, but it's certainly a goal to strive towards.

If something may not be used, then the most optimal thing is to never create it. It never puts pressure on the heap, and never needs to be garbage collected. In TentativeData, a bundle may not have any output at all, and this is unknowable at construction time. So it's cheap enough to check, and construct it on demand, then carry on identically.

On the otherhand, if something is going to be used 100% of the time, it's usually better to provide a New function or an Init method to ensure everything is ready for downstream use. Deciding between those is usually dependant on how many internal fields exist.  eg. The B type is largely manipulated by it's client package via it's fields at the moment, and creating a NewB function for it, while providing compile time safety in what must be provided for initialization, isn't clearly the best approach yet. 

Being able to revisit these choices is possible because the only consumers of these packages are part of the prism runner, so we don't risk breaking unknown users via these changes, as long as the runner works.

As you can see, the right approach is situational. :)

Reply:
Ok, understood. Thank you!